toys/rocks-candy-shop/src/lib.rs

```diff
@@ -0,0 +1,119 @@
+            - add badges to control who can get the collected XRD
+        - allow for different prices for each type of candy
+        - allow candy restocking by a badged account (hint: mutable supply)
+        - tag the candy_vaults with the rri instead of the token symbol
+        - many more...
+*/
+
+blueprint! {
+    struct CandyShop {
+        // The different kinds of candies are kept here with each in a tuple
+        // with a unique tag string that doubles as the candies' token symbol.
+        candy_vaults: Vec<(String,Vault)>
+    }
+
+    impl CandyShop {
+        
+        pub fn new() -> Component {
+            // This constructor sets up an empty candy shop
+            let tagged_vaults: Vec<(String,Vault)> = Vec::new();
+            Self {
+                candy_vaults: tagged_vaults
+            }
+            .instantiate()
+        }
+
+        pub fn initial_supply( supply_size: u32 ) -> Component {
+            // This constructor sets up a variety of candies with the specified amount.
+            let mut tagged_vaults: Vec<(String,Vault)> = Vec::new();
+            // Now define the meta data for each type of candy.
+            let mut metas = vec![("Gumball", "GUM", "The best gumball in the world.")];
+            metas.push(("Jawbreaker", "JAWB", "Jawbreakers teach patience."));
+            metas.push(("Lollipop","LPOP","You can't lick Lollipops!"));
+            metas.push(("Candy Cane", "CANE", "Striped candy rules!"));
+            metas.push(("Jelly Bean", "JELLY", "Jelly Beans are best!"));
+            metas.push(("Mint Candy", "MINT", "Mints are wonderful!"));
+            metas.push(("Gummy Bear", "BEAR", "Gummy Bears rules!"));
+            // Create a supply 
+            for tup in metas {
+                let bucket = ResourceBuilder::new()
+                    .metadata("name", tup.0.to_string())
+                    .metadata("symbol", tup.1.to_string())
+                    .metadata("description", tup.2.to_string())
+                    .new_token_fixed(supply_size);
+                tagged_vaults.push((tup.1.to_string(), Vault::with_bucket(bucket)));
+            }
+            Self {
+                candy_vaults: tagged_vaults
+            }
+            .instantiate()
+        }
+
+        fn take_from_vault(&self, symbol: String, quantity: Decimal) -> Bucket {
+            // private function returns a bucket with the specified number and type of candy (or an empty bucket)
+            for c in &self.candy_vaults[..] {
+                if c.0 == symbol {
+                    let v = &c.1;
+                    if v.amount() >= quantity {
+                        return v.take(quantity)
+                    } else {
+                        break;
+                    }
+                }
+            }
+            let empty_bucket: Bucket = Bucket::new(RADIX_TOKEN); // canonical way to make an empty_bucket
+            return empty_bucket
+        }
```

Comment:
This is an interesting case where the bucket "type" is dynamic and sometimes the method doesn't even want to return a bucket. 

I would consider returning an `Option<Bucket>`, but this introduces complexity from an interface perspective.